File: 567-permutation-in-string/567-permutation-in-string.cs
public class Solution {
    public bool CheckInclusion(string s1, string s2) {
        int len1 = s1.Length, len2 = s2.Length;
        if(len1 > len2)
            return false;
        
        int[] map1 = new int[26];
        for(int i = 0; i < len1; i++){
            map1[s1[i]-97]++;
        }
        
        int[] map2 = new int[26];
        int matchingCount = 0;
        for(int i = 0; i < len1; i++){
            char c = s2[i];
            if(map1[c-97] > 0){
                if(map2[c-97] < map1[c-97]){
                    matchingCount++;
                    if(matchingCount == len1)
                        return true;
                }
                
                 map2[c-97]++;
            }
        }
        
        //sliding window
        int low = 0, hi = len1-1;
        while(hi < len2-1){
            char c = s2[low];
            if(map1[c-97] > 0){
                if(map2[c-97] <= map1[c-97])
                    matchingCount--;
                
                map2[c-97]--;
                
            }
            
            low++;
            
            hi++;
            c = s2[hi];
            if(map1[c-97] > 0){
                if(map2[c-97] < map1[c-97]){
                    matchingCount++;
                }
                
                 map2[c-97]++;
            }
            
            if(matchingCount == len1)
                return true;
            
        }
        
        return matchingCount == len1;
    }
}

File: 567-permutation-in-string/567-permutation-in-string.java
class Solution {
    public boolean checkInclusion(String s1, String s2) {
        int len1 = s1.length();
        int len2 = s2.length();
        
        if(len2 < len1)
            return false;
        
        int[] map1 = new int[26];
        int[] map2 = new int[26];
        
        for(int i = 0; i < len1; i++){
            map1[s1.charAt(i)-97]++;
        }
        
        int matchCount = 0;
        int i = 0;
        for(; i < len1; i++){
            if(map1[s2.charAt(i)-97] > 0){
                map2[s2.charAt(i)-97]++;
                if(map2[s2.charAt(i)-97] <= map1[s2.charAt(i)-97]){
                    matchCount++;
                }
            }
        }
        
        if(matchCount == len1)
            return true;
        
        int low = 0, hi = len1-1;
        //sliding window now
        
        while(hi < len2){
            if(map1[s2.charAt(low)-97] > 0){
                map2[s2.charAt(low)-97]--;
                if(map2[s2.charAt(low)-97] < map1[s2.charAt(low)-97])
                    matchCount--;
            }
            
            low++;
            
            hi++;
            if(hi < len2 && map1[s2.charAt(hi)-97] > 0){
                map2[s2.charAt(hi)-97]++;
                if(map2[s2.charAt(hi)-97] <= map1[s2.charAt(hi)-97]){
                    matchCount++;
                }
            }
            
            if(matchCount == len1)
                return true;
        }
        
        return false;
    }
}